packages/fxa-auth-server/test/local/routes/totp.js

```diff
@@ -741,6 +741,115 @@ describe('totp', () => {
+      );
+      assert.ok(response.qrCodeUrl);
+      assert.ok(response.secret);
+      assert.equal(
+        authServerCacheRedis.set.callCount,
+        1,
+        'stored TOTP token in Redis'
+      );
+    });
+    it('should throw error if the user does not have an existing token', async () => {
+      await setup({
+            db: { email: TEST_EMAIL },
+            totpTokenVerified: false,
+            totpTokenEnabled: false
+          },
+          {},
+          '/totp/replace/start',
+          requestOptions).then(assert.fail, (err) => {
+        assert.deepEqual(err.errno, 220, 'Error number for TOTP does not match');
+        assert.deepEqual(err.message, 'TOTP token does not exist for this account.');
+      });
+    });
+  });
+
+  describe('/totp/replace/confirm', () => {
+    beforeEach(() => {
+      glean.twoFactorAuth.codeComplete.reset();
+    });
+    it('should verify a valid replacement totp code', async () => {
+      const authenticator = new otplib.authenticator.Authenticator();
+      authenticator.options = Object.assign({}, otplib.authenticator.options, {
+        secret,
+      });
+      requestOptions.payload = {
+        code: authenticator.generate(secret),
+      };
+      const response = await setup(
+        {
+          db: { email: TEST_EMAIL },
+          redis: { secret },
+        },
+        {},
+        '/totp/replace/confirm',
+        requestOptions
+      );
+
+      assert.isTrue(response.success);
+      assert.calledOnce(db.replaceTotpToken);
+      assert.calledOnce(authServerCacheRedis.del);
+      assert.calledOnce(glean.twoFactorAuth.codeComplete);
+    });
+
+    it('should fail for an invalid replacement totp code', async () => {
+      await setup(
+        {
+          db: { email: TEST_EMAIL },
+          totpTokenVerified: true,
+          totpTokenEnabled: false,
+          redis: { secret },
+        },
+        {},
+        '/totp/replace/confirm',
+        requestOptions
+      ).then(assert.fail, (err) => {
+        assert.equal(err.message, 'Invalid token confirmation code');
+      });
+    });
+
+    it('should error if session is not verified', async () => {
+      requestOptions.credentials.tokenVerificationId = 'notverified';
+      await setup(
+        { db: { email: TEST_EMAIL } },
+        {},
+        '/totp/replace/confirm',
+        requestOptions
+      ).then(assert.fail, (err) => {
+        assert.equal(err.message, 'Unconfirmed session');
+      });
+    });
+
+    it('should return false if replacement fails', async () => {
+      const authenticator = new otplib.authenticator.Authenticator();
+      authenticator.options = Object.assign({}, otplib.authenticator.options, {
+        secret,
+      });
+      requestOptions.payload = {
+        code: authenticator.generate(secret),
+      };
+      const response = await setup(
+        {
+          db: { email: TEST_EMAIL },
+          redis: { secret }
+        },
+        { replaceTotpToken: true },
```

Comment:
I know your just following the existing patterns, but every time I look at this I have to remember how it works. It doesn't feel intuitive to me.